R/makeTreeSummarizedExperimentFromqiime2.R

```diff
@@ -0,0 +1,245 @@
+#' @noRd
+#' @keywords internal
+read_q2biom <- function(file) {
+    biomobj <- suppressWarnings(biomformat::read_biom(file))
+    feature_tab <- as(biomformat::biom_data(biomobj),"matrix")
+
+    feature_tab
+}
+
+#' Read qiime2 taxa file
+#' @keywords internal
+#' @noRd
+read_q2taxa <- function(file) {
+    taxa <- utils::read.table(file, sep = '\t', header = TRUE)
+    if ("Confidence" %in% names(taxa)) {
+        taxa$Confidence <- NULL
+    }
+
+    as.matrix(taxa)
+}
+
+#' Read qiime2 sample meta data file
+#' @keywords internal
+#' @noRd
+read_q2sample_meta <- function(file) {
+    sam <- utils::read.table(file = file, header = TRUE, sep = "\t", comment.char = "")
+    rownames(sam) <- as.character(sam[, 1])
+
+    # remove row: #q2:types
+    idx <- which(sam == "#q2:types", arr.ind = TRUE)
+    sam <- sam[-idx[, "row"], ]
+
+    S4Vectors::DataFrame(sam)
+}
+
+
+#' Subset taxa according to the taxa in feature table
+#' @keywords internal
+#' @noRd
+subset_taxa_in_feature <- function(taxa_tab, feature_tab) {
+    idx <- match(row.names(feature_tab), taxa_tab[, "Feature.ID"])
+    taxa_tab <- taxa_tab[idx, , drop = FALSE]
+    row.names(taxa_tab) <- taxa_tab[, "Feature.ID"]
+
+    taxa_tab
+}
+
+#' Parse qiime2 taxa in different taxonomic levels
+#' @param taxa `tax_table` object.
+#' @param sep character string containing a regular expression, separator
+#'  between different taxonomic levels, defaults to on compatible with both
+#'  GreenGenes and SILVA `; |;"`.
+#' @param trim_rank_prefix logical whether remove leading characters from
+#'   taxonomic levels, e.g. k__ or D_0__, default `TRUE`.
+#' @return [`phyloseq::taxonomyTable-class`] object.
+#' @keywords internal
+#' @importFrom rlang .data
+#' @noRd
+parse_q2taxonomy <- function(taxa, sep = "; |;", trim_rank_prefix = TRUE) {
+    taxa <- data.frame(taxa)
+    if(trim_rank_prefix){
+        # remove leading characters from GG
+        taxa$Taxon <- gsub("[kpcofgs]__", "", taxa$Taxon)
+        #remove leading characters from SILVA
+        taxa$Taxon <- gsub("D_\\d__", "", taxa$Taxon)
+    }
+
+    taxa <- suppressWarnings(
+        tidyr::separate(
+            taxa, .data$Taxon,
+            c("Kingdom","Phylum","Class","Order","Family","Genus","Species"),
+            sep = sep
+        )
+    )
+    taxa <- sapply(taxa, function(x) ifelse(x == "", NA_character_, x))
+    rownames(taxa) <- taxa[, "Feature.ID"]
+    taxa <- taxa[, setdiff(colnames(taxa), "Feature.ID")]
+
+    taxa
+}
+
+#' check the row.names of feature table is DNA sequence or not
+#' This function is from https://github.com/YuLab-SMU/MicrobiotaProcess/blob/master/R/import_qiime2.R#L169-L177
+#' @keywords internal
+#' @noRd
+is_dna_seq <- function(names){
+    x <- unlist(strsplit(toupper(names[1]), split = ""))
+    freq <- mean(x %in% c("A", "G","T","C", "N", "X", "-"))
+
+    if (length(x) > 30 & freq > 0.9) {
+        return(TRUE)
+    } else {
+        return(FALSE)
+    }
+}
```

Comment:
This is weird.

Just do this:

```
seq <- try({DNAStringSet(seq)}, silent = TRUE)
!is(seq,"try-error")
```

Another thing: If it is a sequence, you are converting it twice, once for the check, once for the actual conversion. Refactor the code to run the conversion only once